text/0000-duration-reform.md

```diff
@@ -0,0 +1,169 @@
+
+This proposal retains the two fields in the existing `Duration`:
+
+* a `u64` of seconds
+* a `u32` of additional nanosecond precision
+
+Timeout APIs defined in terms of milliseconds will truncate `Duration`s that are more than `u32::MAX` in milliseconds, and will reduce the granularity of the nanosecond field.
+
+> A `u32` of milliseconds supports a timeout longer than 45 days.
+
+Future APIs to support a broader set of [Durations][joda-duration] APIs, a [Period][joda-period] and [Instant][joda-instant] type, as well as coercions between these types, would be useful, compatible follow-ups to this RFC.
+
+[joda-duration]: http://www.joda.org/joda-time/key_duration.html
+[joda-period]: http://www.joda.org/joda-time/key_period.html
+[joda-instant]: http://www.joda.org/joda-time/key_instant.html
+
+# Detailed design
+
+A `Duration` represents a period of time represented in terms of nanosecond granularity. It has `u64` seconds and an additional `u32` nanoseconds. There is no concept of a negative `Duration`.
+
+> A negative `Duration` has no meaning for many APIs that may wish
+> to take a `Duration`, which means that all such APIs would need
+> to decide what to do when confronted with a negative `Duration`.
+> As a result, this proposal focuses on the predominant use-cases for
+> `Duration`, where unsigned types remove a number of caveats and
+> ambiguities.
+
+```rust
+pub struct Duration {
+  secs: u64,
+  nanos: u32 // may not be more than 1 billion
+}
+
+impl Duration {
+    /// create a Duration from a number of seconds and an
+    /// additional nanosecond precision. If nanos is one
+    /// billion or greater, it carries into secs.
+    pub fn new(secs: u64, nanos: u32) -> Timeout;
+
+    /// create a Duration from a number of seconds
+    pub fn from_secs(secs: u64) -> Timeout;
+
+    /// create a Duration from a number of milliseconds
+    pub fn from_millis(millis: u64) -> Timeout;
+
+    /// the number of seconds represented by the Duration
+    pub fn secs(self) -> u64;
+
+    /// the number of additional nanosecond precision
+    pub fn nanos(self) -> u32;
+}
+```
+
+When `Duration` is used with a system API that expects `u32` milliseconds, the `Duration`'s precision is coarsened to milliseconds, and, and the number is truncated to `u32::MAX`.
+
+In general, this RFC assumes that timeout APIs permit spurious updates (see, for example, [pthread_cond_timedwait][pthread_cond_timedwait], "Spurious wakeups from the pthread_cond_timedwait() or pthread_cond_wait() functions may occur").
+
+[pthread_cond_timedwait]: http://pubs.opengroup.org/onlinepubs/009695399/functions/pthread_cond_timedwait.html
+
+`Duration` implements:
+
+* `Add`, `Sub`, `Mul`, `Div` which follow the overflow and underflow
+  rules for `u64` when applied to the `secs` field (in particular,
+  `Sub` will panic if the result would be negative). Nanoseconds
+  must be less than 1 billion and great than or equal to 0, and carry 
+  into the `secs` field.
```

Comment:
Will it follow the rules of `u64` or will it also panic in non-debug mode?

I'd vote for the latter, because these duration-manipulating methods are unlikely to be called in performance-sensitive code.